scalyr_agent/line_matcher.py

```diff
@@ -125,6 +169,135 @@ def _readline( self, file_like, max_length=0 ):
+        LineMatcher.__init__( self, max_line_length, line_completion_wait_time )
+
+        self._json_message_field = json_message_field
+        self._json_timestamp_field= json_timestamp_field
+
+    def _truncated_log_line( self, line, max_length ):
+        """
+        Truncates a line if it is greater than `max_length`, and appends a message saying the number of bytes truncted
+        @return: A LogLine with the truncated message
+        @rtype: LogLine
+        """
+
+        line_len = len(line)
+        if line_len > max_length:
+            line = line[:max_length]
+            line += "...truncated %d bytes..." % (line_len - max_length)
+        return LogLine( line )
+
+    def _readline( self, file_like, max_length=0, current_file=None ):
+        """ Takes a file_like object (e.g. anything conforming to python's file interface
+        and attempts to parse the line as JSON.
+
+        This method will attempt to process the line as json, extracting a log message, a timestamp and any
+        other remaining fields, all of which are returned as a LogLine object.
+
+        When parsing as json, each line is required to be a fully formed json object, otherwise the full contents of
+        the line will be returned unprocessed.
+
+        The configuration options 'json_message_field' and 'json_timestamp_field' are used to specify which field to use
+        as the log message, and which field to use for the timestamp.
+
+        'json_message_field' defaults to 'log' and if no field with this name is found, the function will return the
+        full line unprocessed.
+
+        'json_timestamp_field' defaults to 'time', and the value of this field is required to be in rfc3339 format
+        otherwise an error will occur.  If the field does not exist in the json object, then the agent will use
+        the current time instead.  Note, the value specified in the timestamp field might not be the final value uploaded
+        to the server, as the agent ensures that the timestamps of all messages are monotonically increasing.  For
+        this reason, if the timestamp field is found, a 'raw_timestamp' attributed is also added to the LogLine's attrs.
+
+        All other fields of the json object will be stored in the attrs dict of the LogLine object.
+
+        The method will read up to `read_page_size` bytes from the buffer, and truncate any log message to `max_line_size` bytes.
+
+        @param file_like: a file like object (e.g. StringIO)
+        @param max_length: the maximum length to read from the file_like object
+
+        @return: a tuple containing the content read from the file_like object and whether
+            this is a partial line or not.  `partial` will only be true if `read_page_size`
+            was exceeded and a full json object wasn't found.
+
+        @rtype: (LogLine, bool)
+        """
+
+        # set some sensible defaults
+        if max_length == 0:
```

Comment:
I think use `-1` or `None` to signal no `max_length` was given.  `0` is too close to something that could be passed in.